sdlib/d/gc/tcache.d

```diff
@@ -320,6 +225,92 @@ public:
 
 private:
 
+	/**
+	 * Appendable's mechanics:
+	 * 
+	 *  __data__  _____free space_______
+	 * /        \/                      \
+	 * -----sss s....... ....... ........
+	 *      \___________________________/
+	 * 	           Capacity is 27
+	 * 
+	 * If the slice's end doesn't match the used capacity,
+	 * then we return 0 in order to force a reallocation
+	 * when appending:
+	 * 
+	 *  ___data____  ____free space_____
+	 * /           \/                   \
+	 * -----sss s---.... ....... ........
+	 *      \___________________________/
+	 * 	           Capacity is 0
+	 * 
+	 * See also: https://dlang.org/spec/arrays.html#capacity-reserve
+	 */
+
+	bool getOrExtendCapacity(const void[] slice, size_t size,
+	                         ref size_t sliceCapacity) {
+		auto pd = maybeGetPageDescriptor(slice.ptr);
+		auto e = pd.extent;
+		if (e is null) {
+			sliceCapacity = 0;
+			return false;
+		}
+
+		size_t usedCapacity;
+
+		size_t getSliceCapacity(const void[] space) {
+			// Slice must not end before valid data ends, or capacity is zero.
+			// To be appendable, the slice end must match the alloc's used
+			// capacity, and the latter may not be zero.
+			auto startIndex = slice.ptr - space.ptr;
+			auto stopIndex = startIndex + slice.length;
+
+			if (stopIndex == 0 || stopIndex != usedCapacity) {
+				return 0;
+			}
+
+			return space.length - startIndex;
+		}
+
+		if (pd.isSlab()) {
+			import d.gc.slab;
+			auto sg = SlabAllocGeometry(pd, cast(void*) slice.ptr);
```

Comment:
Why the cast?

Reply:
Slab alloc geometry doesn't modify anything, so if that's a problem can't it be made const?